### ISSUE TITLE
eval_shape does not work with make_array_from_callaback

### ISSUE THREAD

**Issue by namiller** [google/jax#19542]
### Description

eval_shape does not work with make_array_from_callaback when the generated array is not fully replicated. This doesn't seem like a fundamental limitation since the returned ShapeDtypeStruct has a sharding parameter that could be correctly populated.

Simple Repro:
```

import jax

import numpy as np
import jax.numpy as jnp


def make_gda(shape_spec):
  local_array = np.zeros(shape_spec.shape)
  global_shape = (len(local_array) * jax.process_count(), *local_array.shape[1:])
  mesh = jax.sharding.Mesh(np.array(jax.devices()).reshape(jax.process_count(), -1), ('batch', 'other'))
  sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec('batch', 'other'))
  return jax.make_array_from_callback(global_shape, sharding, lambda x: local_array)

jax.eval_shape(make_gda, jax.ShapeDtypeStruct(shape=(10, 8), dtype=jnp.bfloat16))

```

### Which accelerator(s) are you using?

TPU

**Comment by jakevdp** [comment]
Hi - I think this is intended behavior: `make_array_from_callback` is not meant for use in traced contexts. @yashk2810 may be able to provide more background.

**Comment by yashk2810** [comment]
The use case you have doesn't make sense to me. You already have the shape, dtype and sharding information. Why are you running eval_shape on it?

**Comment by namiller** [comment]
Yeah, that was a contrived example to demonstrate the failure mode. In practice I run into this because I have a method that does some setup/initialization logic and takes in a shape spec for the dataset element (and subsequently calls a flax module's init in a jitted context to construct a sharded parameters object for it). During normal operation this method gets called in a jitted context, during checkpoint loading it gets called through eval_shape.

There are obviously ways to avoid this, however this still seems like something that doesn't work and which:

1) has an obvious/well defined behavior that could allow it to work
2) does not produce a reasonable error message when hit
3) does not have it's current behavior documented on either eval_shape or make_array_from_callback as far as I can tell.

**Comment by jakevdp** [comment]
Points 2 and 3 should be addressed by https://github.com/google/jax/pull/19580. But I agree on 1. Yash and I have been discussing off-thread whether this would make sense.
